File: python/x402/tests/integrations/_smart_accounts.py
"""Integration-test helpers for real smart-account signature wrapping.

Coinbase Smart Wallet (ERC-4337): replay-safe EIP-712 + SignatureWrapper.
Biconomy Nexus (ERC-7579): ERC-7739 nested EIP-712 + validator prefix.
"""

from __future__ import annotations

import struct
from typing import Any

from eth_abi import encode
from eth_account import Account
from eth_account.messages import encode_typed_data
from eth_utils import to_checksum_address
from web3 import Web3

from x402.mechanisms.evm.constants import EIP1271_MAGIC_VALUE
from x402.mechanisms.evm.types import TypedDataDomain, TypedDataField
from x402.mechanisms.evm.verify import _hash_typed_data

COINBASE_SMART_WALLET_FACTORY = "0xBA5ED110eFDBa3D005bfC882d75358ACBbB85842"
NEXUS_ACCOUNT_FACTORY = "0x000000002c9A405a196f2dc766F2476B731693c3"
NEXUS_BOOTSTRAP = "0x000000007BfEdA33ac982cb38eAaEf5D7bCC954c"
NEXUS_K1_VALIDATOR = "0x0000000002d3cC5642A748B6783F32C032616E03"
NEXUS_DEFAULT_VALIDATOR_PREFIX = bytes(20)

EIP712_DOMAIN_ABI = [
    {
        "name": "eip712Domain",
        "type": "function",
        "stateMutability": "view",
        "inputs": [],
        "outputs": [
            {"name": "fields", "type": "bytes1"},
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
            {"name": "salt", "type": "bytes32"},
            {"name": "extensions", "type": "uint256[]"},
        ],
    }
]

IS_VALID_SIGNATURE_ABI = [
    {
        "name": "isValidSignature",
        "type": "function",
        "stateMutability": "view",
        "inputs": [
            {"name": "hash", "type": "bytes32"},
            {"name": "signature", "type": "bytes"},
        ],
        "outputs": [{"type": "bytes4"}],
    }
]


def _domain_dict(domain: TypedDataDomain | dict[str, Any]) -> dict[str, Any]:
    if isinstance(domain, TypedDataDomain):
        return {
            "name": domain.name,
            "version": domain.version,
            "chainId": domain.chain_id,
            "verifyingContract": domain.verifying_contract,
        }
    return domain


def _types_dict(
    types: dict[str, list[TypedDataField | dict[str, str]]],
) -> dict[str, list[dict[str, str]]]:
    out: dict[str, list[dict[str, str]]] = {}
    for type_name, fields in types.items():
        out[type_name] = [
            {"name": f.name, "type": f.type} if isinstance(f, TypedDataField) else f for f in fields
        ]
    return out


def _prepare_message_for_eth_account(message: dict[str, Any]) -> dict[str, Any]:
    msg = dict(message)
    if "nonce" in msg and isinstance(msg["nonce"], bytes):
        msg["nonce"] = "0x" + msg["nonce"].hex()
    return msg


def hash_typed_data_digest(
    domain: TypedDataDomain | dict[str, Any],
    types: dict[str, list[TypedDataField | dict[str, str]]],
    primary_type: str,
    message: dict[str, Any],
) -> bytes:
    return _hash_typed_data(domain, types, primary_type, message)


def wrap_coinbase_signature(signature: bytes, owner_index: int = 0) -> bytes:
    if len(signature) != 65:
        raise ValueError(f"expected 65-byte signature, got {len(signature)}")
    # Coinbase Smart Wallet expects raw r||s||v bytes, not ABI-encoded (bytes32,bytes32,uint8).
    return encode(["uint8", "bytes"], [owner_index, signature])


def wrap_erc7739_typed_data_signature(
    domain: TypedDataDomain | dict[str, Any],
    types: dict[str, list[TypedDataField | dict[str, str]]],
    primary_type: str,
    message: dict[str, Any],
    signature: bytes,
) -> bytes:
    domain_dict = _domain_dict(domain)
    types_dict = _types_dict(types)
    full_message = {
        "types": types_dict,
        "primaryType": primary_type,
        "domain": domain_dict,
        "message": _prepare_message_for_eth_account(message),
    }
    signable = encode_typed_data(full_message=full_message)
    encoded_type = _encode_type(primary_type, types_dict)
    type_bytes = encoded_type.encode("utf-8")
    return (
        signature
        + signable.header
        + signable.body
        + type_bytes
        + struct.pack(">H", len(type_bytes))
    )


def _encode_type(primary_type: str, types: dict[str, list[dict[str, str]]]) -> str:
    deps: list[str] = []
    seen: set[str] = set()

    def collect(type_name: str) -> None:
        if type_name in seen or type_name == "EIP712Domain":
            return
        seen.add(type_name)
        for field in types.get(type_name, []):
            base = field["type"].split("[", 1)[0]
            if base in types and base != "EIP712Domain":
                collect(base)
        if type_name != "EIP712Domain":
            deps.append(type_name)

    collect(primary_type)
    parts: list[str] = []
    for dep in deps:
        fields = types[dep]
        inner = ",".join(f"{f['type']} {f['name']}" for f in fields)
        parts.append(f"{dep}({inner})")
    return "".join(parts)


def sign_coinbase_smart_wallet_typed_data(
    owner: Account,
    smart_account_address: str,
    domain: TypedDataDomain | dict[str, Any],
    types: dict[str, list[TypedDataField | dict[str, str]]],
    primary_type: str,
    message: dict[str, Any],
    chain_id: int = 84532,
) -> bytes:
    original_hash = hash_typed_data_digest(domain, types, primary_type, message)
    replay_message = {
        "types": {
            "EIP712Domain": [
                {"name": "name", "type": "string"},
                {"name": "version", "type": "string"},
                {"name": "chainId", "type": "uint256"},
                {"name": "verifyingContract", "type": "address"},
            ],
            "CoinbaseSmartWalletMessage": [{"name": "hash", "type": "bytes32"}],
        },
        "primaryType": "CoinbaseSmartWalletMessage",
        "domain": {
            "name": "Coinbase Smart Wallet",
            "version": "1",
            "chainId": chain_id,
            "verifyingContract": to_checksum_address(smart_account_address),
        },
        "message": {"hash": "0x" + original_hash.hex()},
    }
    signable = encode_typed_data(full_message=replay_message)
    inner_sig = owner.sign_message(signable).signature
    return wrap_coinbase_signature(inner_sig)


def fetch_nexus_verifier_domain(w3: Web3, nexus_address: str) -> dict[str, Any]:
    contract = w3.eth.contract(
        address=Web3.to_checksum_address(nexus_address),
        abi=EIP712_DOMAIN_ABI,
    )
    result = contract.functions.eip712Domain().call()
    return {
        "name": result[1],
        "version": result[2],
        "chainId": result[3],
        "verifyingContract": to_checksum_address(result[4]),
        "salt": result[5],
    }


def sign_nexus_typed_data(
    owner: Account,
    nexus_address: str,
    validator_address: str,
    domain: TypedDataDomain | dict[str, Any],
    types: dict[str, list[TypedDataField | dict[str, str]]],
    primary_type: str,
    message: dict[str, Any],
    w3: Web3,
) -> bytes:
    verifier_domain = fetch_nexus_verifier_domain(w3, nexus_address)
    types_dict = _types_dict(types)
    types_dict["TypedDataSign"] = [
        {"name": "contents", "type": primary_type},
        {"name": "name", "type": "string"},
        {"name": "version", "type": "string"},
        {"name": "chainId", "type": "uint256"},
        {"name": "verifyingContract", "type": "address"},
        {"name": "salt", "type": "bytes32"},
    ]
    nested_message = {
        "types": types_dict,
        "primaryType": "TypedDataSign",
        "domain": _domain_dict(domain),
        "message": {
            "contents": _prepare_message_for_eth_account(message),
            "name": verifier_domain["name"],
            "version": verifier_domain["version"],
            "chainId": verifier_domain["chainId"],
            "verifyingContract": verifier_domain["verifyingContract"],
            "salt": verifier_domain["salt"],
        },
    }
    signable = encode_typed_data(full_message=nested_message)
    inner_sig = owner.sign_message(signable).signature
    wrapped = wrap_erc7739_typed_data_signature(domain, types, primary_type, message, inner_sig)
    return NEXUS_DEFAULT_VALIDATOR_PREFIX + wrapped


def verify_is_valid_signature(
    w3: Web3,
    account_address: str,
    digest: bytes,
    signature: bytes,
) -> bool:
    contract = w3.eth.contract(
        address=Web3.to_checksum_address(account_address),
        abi=IS_VALID_SIGNATURE_ABI,
    )
    result = contract.functions.isValidSignature(digest, signature).call()
    if isinstance(result, (bytes, bytearray)):
        magic = "0x" + result[:4].hex()
    else:
        magic = result
    return str(magic).lower() == EIP1271_MAGIC_VALUE.lower()


class CoinbaseSmartWalletSigner:
    """Client signer presenting a Coinbase Smart Wallet address with wrapped signatures."""

    def __init__(self, owner: Account, smart_account_address: str, chain_id: int = 84532) -> None:
        self._owner = owner
        self._address = to_checksum_address(smart_account_address)
        self._chain_id = chain_id

    @property
    def address(self) -> str:
        return self._address

    def sign_typed_data(
        self,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
    ) -> bytes:
        return sign_coinbase_smart_wallet_typed_data(
            self._owner,
            self._address,
            domain,
            types,
            primary_type,
            message,
            self._chain_id,
        )


class NexusSmartAccountSigner:
    """Client signer presenting a Biconomy Nexus address with ERC-7739 wrapped signatures."""

    def __init__(
        self,
        owner: Account,
        nexus_address: str,
        w3: Web3,
        validator_address: str = NEXUS_K1_VALIDATOR,
    ) -> None:
        self._owner = owner
        self._address = to_checksum_address(nexus_address)
        self._w3 = w3
        self._validator = validator_address

    @property
    def address(self) -> str:
        return self._address

    def sign_typed_data(
        self,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
    ) -> bytes:
        return sign_nexus_typed_data(
            self._owner,
            self._address,
            self._validator,
            domain,
            types,
            primary_type,
            message,
            self._w3,
        )

File: typescript/packages/mechanisms/evm/test/integrations/helpers/smartAccounts.ts
/**
 * Integration-test helpers for real smart-account signature wrapping.
 *
 * Coinbase Smart Wallet (ERC-4337): replay-safe EIP-712 + SignatureWrapper.
 * Biconomy Nexus (ERC-7579): ERC-7739 nested EIP-712 + validator prefix.
 */

import {
  concat,
  createPublicClient,
  createWalletClient,
  encodeAbiParameters,
  encodeFunctionData,
  encodePacked,
  getAddress,
  hashTypedData,
  http,
  keccak256,
  parseAbiParameters,
  sliceHex,
  type Hex,
  type TypedDataDefinition,
} from "viem";
import { privateKeyToAccount, type PrivateKeyAccount } from "viem/accounts";
import { baseSepolia } from "viem/chains";
import { wrapTypedDataSignature } from "viem/experimental/erc7739";
import type { ClientEvmSigner } from "../../../../src/signer";

/** Coinbase Smart Wallet Factory v1.1 (Base Sepolia). */
export const COINBASE_SMART_WALLET_FACTORY = "0xBA5ED110eFDBa3D005bfC882d75358ACBbB85842" as const;

/** Biconomy Nexus Account Factory (MEE v1.3.1, Base Sepolia). */
export const NEXUS_ACCOUNT_FACTORY = "0x000000002c9A405a196f2dc766F2476B731693c3" as const;

/** Biconomy Nexus Bootstrap (MEE v1.3.0, Base Sepolia). */
export const NEXUS_BOOTSTRAP = "0x000000007BfEdA33ac982cb38eAaEf5D7bCC954c" as const;

/** MEE K1 validator / default validator module (Base Sepolia). */
export const NEXUS_K1_VALIDATOR = "0x0000000002d3cC5642A748B6783F32C032616E03" as const;

/**
 * ERC-7579 signature prefix for accounts initialized with initNexusWithDefaultValidator.
 * Nexus routes address(0) to the bootstrap default validator (see _handleValidator).
 */
export const NEXUS_DEFAULT_VALIDATOR_PREFIX = "0x0000000000000000000000000000000000000000" as const;

export const EIP1271_MAGIC = "0x1626ba7e" as const;

const SIGNATURE_WRAPPER_STRUCT = {
  components: [
    { name: "ownerIndex", type: "uint8" },
    { name: "signatureData", type: "bytes" },
  ],
  name: "SignatureWrapper",
  type: "tuple",
} as const;

const COINBASE_FACTORY_ABI = [
  {
    name: "createAccount",
    type: "function",
    stateMutability: "payable",
    inputs: [
      { name: "owners", type: "bytes[]" },
      { name: "nonce", type: "uint256" },
    ],
    outputs: [{ type: "address" }],
  },
  {
    name: "getAddress",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "owners", type: "bytes[]" },
      { name: "nonce", type: "uint256" },
    ],
    outputs: [{ type: "address" }],
  },
] as const;

const NEXUS_FACTORY_ABI = [
  {
    name: "createAccount",
    type: "function",
    stateMutability: "payable",
    inputs: [
      { name: "initData", type: "bytes" },
      { name: "salt", type: "bytes32" },
    ],
    outputs: [{ type: "address" }],
  },
  {
    name: "computeAccountAddress",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "initData", type: "bytes" },
      { name: "salt", type: "bytes32" },
    ],
    outputs: [{ type: "address" }],
  },
] as const;

const NEXUS_BOOTSTRAP_ABI = [
  {
    name: "initNexusWithDefaultValidator",
    type: "function",
    stateMutability: "payable",
    inputs: [{ name: "data", type: "bytes" }],
    outputs: [],
  },
] as const;

/** Build Nexus initializeAccount initData for a single EOA owner (default K1 validator). */
export function buildNexusInitData(owner: `0x${string}`): Hex {
  const ownerData = encodePacked(["address"], [getAddress(owner)]);
  const bootstrapCall = encodeFunctionData({
    abi: NEXUS_BOOTSTRAP_ABI,
    functionName: "initNexusWithDefaultValidator",
    args: [ownerData],
  });
  return encodeAbiParameters(parseAbiParameters("address bootstrap, bytes bootstrapCall"), [
    getAddress(NEXUS_BOOTSTRAP),
    bootstrapCall,
  ]);
}

/** Deterministic salt for owner + index (matches integration-test deployment). */
export function nexusAccountSalt(owner: `0x${string}`, index = 0n): Hex {
  return keccak256(encodePacked(["address", "uint256"], [getAddress(owner), index]));
}

const EIP712_DOMAIN_ABI = [
  {
    name: "eip712Domain",
    type: "function",
    stateMutability: "view",
    inputs: [],
    outputs: [
      { name: "fields", type: "bytes1" },
      { name: "name", type: "string" },
      { name: "version", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" },
      { name: "salt", type: "bytes32" },
      { name: "extensions", type: "uint256[]" },
    ],
  },
] as const;

const IS_VALID_SIGNATURE_ABI = [
  {
    name: "isValidSignature",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "hash", type: "bytes32" },
      { name: "signature", type: "bytes" },
    ],
    outputs: [{ type: "bytes4" }],
  },
] as const;

/** Encode a single EOA owner for CoinbaseSmartWalletFactory.createAccount. */
export function encodeCoinbaseOwner(owner: `0x${string}`): `0x${string}` {
  return encodeAbiParameters(parseAbiParameters("address"), [getAddress(owner)]);
}

/** Predict Coinbase Smart Wallet address for an owner + nonce. */
export async function predictCoinbaseSmartWalletAddress(
  owner: `0x${string}`,
  nonce = 0n,
  rpcUrl = "https://sepolia.base.org",
): Promise<`0x${string}`> {
  const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
  const owners = [encodeCoinbaseOwner(owner)];
  return pc.readContract({
    address: COINBASE_SMART_WALLET_FACTORY,
    abi: COINBASE_FACTORY_ABI,
    functionName: "getAddress",
    args: [owners, nonce],
  });
}

/** Deploy Coinbase Smart Wallet; returns the account address. */
export async function deployCoinbaseSmartWallet(
  deployerKey: `0x${string}`,
  owner: `0x${string}`,
  nonce = 0n,
  rpcUrl = "https://sepolia.base.org",
): Promise<`0x${string}`> {
  const deployer = privateKeyToAccount(deployerKey);
  const wc = createWalletClient({ account: deployer, chain: baseSepolia, transport: http(rpcUrl) });
  const owners = [encodeCoinbaseOwner(owner)];
  return wc
    .writeContract({
      address: COINBASE_SMART_WALLET_FACTORY,
      abi: COINBASE_FACTORY_ABI,
      functionName: "createAccount",
      args: [owners, nonce],
    })
    .then(async hash => {
      const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
      await pc.waitForTransactionReceipt({ hash });
      return predictCoinbaseSmartWalletAddress(owner, nonce, rpcUrl);
    });
}

/** Predict Biconomy Nexus address for an owner + index. */
export async function predictNexusAccountAddress(
  owner: `0x${string}`,
  index = 0n,
  rpcUrl = "https://sepolia.base.org",
): Promise<`0x${string}`> {
  const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
  const initData = buildNexusInitData(owner);
  const salt = nexusAccountSalt(owner, index);
  return pc.readContract({
    address: NEXUS_ACCOUNT_FACTORY,
    abi: NEXUS_FACTORY_ABI,
    functionName: "computeAccountAddress",
    args: [initData, salt],
  });
}

/** Deploy Biconomy Nexus account; returns the account address. */
export async function deployNexusAccount(
  deployerKey: `0x${string}`,
  owner: `0x${string}`,
  index = 0n,
  rpcUrl = "https://sepolia.base.org",
): Promise<`0x${string}`> {
  const deployer = privateKeyToAccount(deployerKey);
  const wc = createWalletClient({ account: deployer, chain: baseSepolia, transport: http(rpcUrl) });
  const initData = buildNexusInitData(owner);
  const salt = nexusAccountSalt(owner, index);
  return wc
    .writeContract({
      address: NEXUS_ACCOUNT_FACTORY,
      abi: NEXUS_FACTORY_ABI,
      functionName: "createAccount",
      args: [initData, salt],
    })
    .then(async hash => {
      const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
      await pc.waitForTransactionReceipt({ hash });
      return predictNexusAccountAddress(owner, index, rpcUrl);
    });
}

/** Wrap a 65-byte ECDSA signature in Coinbase Smart Wallet SignatureWrapper format. */
export function wrapCoinbaseSmartWalletSignature(signatureHex: Hex, ownerIndex = 0): Hex {
  const r = sliceHex(signatureHex, 0, 32);
  const s = sliceHex(signatureHex, 32, 64);
  const v = Number(`0x${signatureHex.slice(130, 132)}`);
  const signatureData = encodePacked(["bytes32", "bytes32", "uint8"], [r, s, v]);
  return encodeAbiParameters([SIGNATURE_WRAPPER_STRUCT], [{ ownerIndex, signatureData }]);
}

/** Create replay-safe typed data for Coinbase Smart Wallet signing. */
export function createCoinbaseReplaySafeTypedData(
  typedData: TypedDataDefinition,
  smartAccountAddress: `0x${string}`,
  chainId: number | bigint,
): TypedDataDefinition {
  const originalHash = hashTypedData(typedData);
  return {
    domain: {
      name: "Coinbase Smart Wallet",
      version: "1",
      chainId: Number(chainId),
      verifyingContract: getAddress(smartAccountAddress),
    },
    types: {
      CoinbaseSmartWalletMessage: [{ name: "hash", type: "bytes32" }],
    },
    primaryType: "CoinbaseSmartWalletMessage",
    message: { hash: originalHash },
  };
}

/** Sign EIP-712 typed data for Coinbase Smart Wallet (replay-safe + SignatureWrapper). */
export async function signCoinbaseSmartWalletTypedData(
  ownerAccount: PrivateKeyAccount,
  smartAccountAddress: `0x${string}`,
  typedData: TypedDataDefinition,
): Promise<Hex> {
  const chainId =
    typeof typedData.domain?.chainId === "bigint"
      ? typedData.domain.chainId
      : BigInt(typedData.domain?.chainId ?? baseSepolia.id);
  const replaySafe = createCoinbaseReplaySafeTypedData(typedData, smartAccountAddress, chainId);
  const innerSig = await ownerAccount.signTypedData(replaySafe as never);
  return wrapCoinbaseSmartWalletSignature(innerSig);
}

/** Fetch Nexus eip712Domain for ERC-7739 verifierDomain. */
export async function fetchNexusVerifierDomain(
  accountAddress: `0x${string}`,
  rpcUrl = "https://sepolia.base.org",
) {
  const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
  const result = await pc.readContract({
    address: accountAddress,
    abi: EIP712_DOMAIN_ABI,
    functionName: "eip712Domain",
  });
  return {
    name: result[1],
    version: result[2],
    chainId: result[3],
    verifyingContract: getAddress(result[4] as `0x${string}`),
    salt: result[5],
  };
}

/** Sign EIP-712 typed data for Biconomy Nexus (ERC-7739 + validator prefix). */
export async function signNexusTypedData(
  ownerAccount: PrivateKeyAccount,
  nexusAddress: `0x${string}`,
  _validatorAddress: `0x${string}`,
  typedData: TypedDataDefinition,
  rpcUrl = "https://sepolia.base.org",
): Promise<Hex> {
  const verifierDomain = await fetchNexusVerifierDomain(nexusAddress, rpcUrl);
  const nestedSig = await ownerAccount.signTypedData({
    domain: typedData.domain,
    types: {
      ...(typedData.types as Record<string, unknown>),
      TypedDataSign: [
        { name: "contents", type: typedData.primaryType as string },
        { name: "name", type: "string" },
        { name: "version", type: "string" },
        { name: "chainId", type: "uint256" },
        { name: "verifyingContract", type: "address" },
        { name: "salt", type: "bytes32" },
      ],
    },
    primaryType: "TypedDataSign",
    message: {
      contents: typedData.message,
      name: verifierDomain.name,
      version: verifierDomain.version,
      chainId: verifierDomain.chainId,
      verifyingContract: verifierDomain.verifyingContract,
      salt: verifierDomain.salt,
    },
  } as never);
  const wrapped = wrapTypedDataSignature({
    ...typedData,
    signature: nestedSig,
  } as never);
  return concat([NEXUS_DEFAULT_VALIDATOR_PREFIX, wrapped]);
}

/** Verify wrapped signature via on-chain isValidSignature eth_call. */
export async function verifyIsValidSignature(
  accountAddress: `0x${string}`,
  digest: Hex,
  signature: Hex,
  rpcUrl = "https://sepolia.base.org",
): Promise<boolean> {
  const pc = createPublicClient({ chain: baseSepolia, transport: http(rpcUrl) });
  const result = await pc.readContract({
    address: accountAddress,
    abi: IS_VALID_SIGNATURE_ABI,
    functionName: "isValidSignature",
    args: [digest, signature],
  });
  return result.toLowerCase() === EIP1271_MAGIC;
}

/** ClientEvmSigner that presents a smart-account address and wraps signatures for Coinbase Smart Wallet. */
export function createCoinbaseSmartWalletClientSigner(
  ownerAccount: PrivateKeyAccount,
  smartAccountAddress: `0x${string}`,
): ClientEvmSigner {
  return {
    address: getAddress(smartAccountAddress),
    signTypedData: async typedData =>
      signCoinbaseSmartWalletTypedData(ownerAccount, smartAccountAddress, typedData),
  };
}

/** ClientEvmSigner that presents a Nexus address and wraps signatures with ERC-7739 + validator prefix. */
export function createNexusClientSigner(
  ownerAccount: PrivateKeyAccount,
  nexusAddress: `0x${string}`,
  validatorAddress: `0x${string}` = NEXUS_K1_VALIDATOR,
  rpcUrl = "https://sepolia.base.org",
): ClientEvmSigner {
  return {
    address: getAddress(nexusAddress),
    signTypedData: async typedData =>
      signNexusTypedData(ownerAccount, nexusAddress, validatorAddress, typedData, rpcUrl),
  };
}
